format-specs/geoparquet.md

```diff
@@ -83,10 +81,75 @@ The optional `epoch` field allows to specify this in case the `crs` field define
 
 Note that the current version of the spec only allows for a subset of WKB: 2D or 3D geometries of the standard geometry types (the Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon, and GeometryCollection geometry types). This means that M values or non-linear geometry types are not yet supported.
 
+WKB geometry columns MUST be stored using the `BYTE_ARRAY` parquet type.
+
+Implementation note: when using WKB encoding with the ecosystem of Arrow libraries, Parquet types such as `BYTE_ARRAY` might not be directly accessible. Instead, the corresponding Arrow data type can be `Arrow::Type::BINARY` (for arrays that whose elements can be indexed through a 32-bit index) or `Arrow::Type::LARGE_BINARY` (64-bit index). It is recommended that GeoParquet readers are compatible with both data types, and writers preferably use `Arrow::Type::BINARY` (thus limiting to row groups with content smaller than 2 GB) for larger compatibility.
+
+##### Native encodings (based on GeoArrow)
+
+Using the single-geometry type encodings (i.e., `"point"`, `"linestring"`, `"polygon"`, `"multipoint"`, `"multilinestring"`, `"multipolygon"`) may provide better performance and enable readers to leverage more features of the Parquet format to accelerate geospatial queries (e.g., row group-level min/max statistics). These encodings correspond to extension name suffix in the [GeoArrow metadata specification for extension names](https://geoarrow.org/extension-types#extension-names) to signify the memory layout used by the geometry column. GeoParquet uses the separated (struct) representation of coordinates for single-geometry type encodings because this encoding results in useful column statistics when row groups and/or files contain related features.
+
+The actual coordinates of the geometries MUST be stored as native numbers, i.e. using
+the `DOUBLE` parquet type in a (repeated) group of fields (exact repetition depending
+on the geometry type).
+
+For the `"point"` geometry type, this results in a struct of two fields for x
+and y coordinates (in case of 2D geometries):
+
+```
+// "point" geometry column as simple field with two child fields for x and y
+optional group geometry {
+  required double x;
+  required double y;
+}
+```
+
+For the other geometry types, those x and y coordinate values MUST be embedded
+in repeated groups (`LIST` logical parquet type). For example, for the
+`"multipolygon"` geometry type:
+
+```
+// "multipolygon" geometry column with multiple levels of nesting
+optional group geometry (List) {
+  // the parts of the MultiPolygon
+  repeated group list {
```

Comment:
I mentioned this issue above at https://github.com/opengeospatial/geoparquet/pull/189#discussion_r1520080717

The Parquet spec specifies that those names should be "list" and "element": https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists 

But it also mentions that for backwards compatibility you should be able to read lists with other names:

> It is required that the repeated group of elements is named `list` and that its element field is named `element`. However, these names may not be used in existing data and should not be enforced as errors when reading.

Until recently, PyArrow / Arrow C++ actually created such "non-compliant" files, but there is a flag to control this, for which the default has been changed to create compliant files by default since pyarrow 13.0 (https://github.com/apache/arrow/issues/29781).

Illustrating this:

```python
point_type = pa.struct([pa.field("x", pa.float64(), nullable=False), pa.field("y", pa.float64(), nullable=False)])
polygon_type = pa.list_(
    pa.field("rings", pa.list_(pa.field("vertices", point_type, nullable=False)), nullable=False)
)
multipolygon_type = pa.list_(pa.field("polygons", _polygon_type, nullable=False))

table = pa.table({"geometry": pa.array([[[[{'x': 1.0, 'y': 2.0}, {'x': 2.0, 'y': 3.0}]]]], type=multipolygon_type)})

>>> pq.write_table(table, "test_multipolygon_non_nullable-compliant.parquet")
>>> pq.read_metadata("test_multipolygon_non_nullable-compliant.parquet").schema
<pyarrow._parquet.ParquetSchema object at 0x7efda2333b40>
required group field_id=-1 schema {
  optional group field_id=-1 geometry (List) {
    repeated group field_id=-1 list {
      required group field_id=-1 element (List) {
        repeated group field_id=-1 list {
          required group field_id=-1 element (List) {
            repeated group field_id=-1 list {
              required group field_id=-1 element {
                required double field_id=-1 x;
                required double field_id=-1 y;
              }
            }
          }
        }
      }
    }
  }
}

>>> pq.write_table(table, "test_multipolygon_non_nullable-non_compliant.parquet", use_compliant_nested_type=False)
>>> pq.read_metadata("test_multipolygon_non_nullable-non_compliant.parquet").schema
<pyarrow._parquet.ParquetSchema object at 0x7efda135fd00>
required group field_id=-1 schema {
  optional group field_id=-1 geometry (List) {
    repeated group field_id=-1 list {
      required group field_id=-1 polygons (List) {
        repeated group field_id=-1 list {
          required group field_id=-1 rings (List) {
            repeated group field_id=-1 list {
              required group field_id=-1 vertices {
                required double field_id=-1 x;
                required double field_id=-1 y;
              }
            }
          }
        }
      }
    }
  }
}
```

Specifically for Arrow C++ (or other implementation that do this), we do store the original Arrow schema in the Parquet metadata, and so we could preserve the custom names on roundtrip. But it seems we are not doing that currently.

But this actually raises the question: for the GeoArrow spec itself, how "required" are those custom field names? Are you supposed to rename them after reading a compliant Parquet file that will not have those names? 
At the moment the GeoArrow spec itself seems to mention "should" for the names (https://github.com/geoarrow/geoarrow/blob/main/format.md#memory-layouts)

> Thinking that for bounding box stats I'll need to find the leaf arrays by name and for multipolygon it'll be something like `geometry.list.list.list.x` and ``geometry.list.list.list.y`?

To get back to that initial question: it will be something like `geometry.list.element.list.element.x` (with the number of `list.element` repetitions depending on the geometry type)

Reply:
> But it also mentions that for backwards compatibility you should be able to read lists with other names:

👍 

> But this actually raises the question: for the GeoArrow spec itself, how "required" are those custom field names? Are you supposed to rename them after reading a compliant Parquet file that will not have those names? 

My implementation will likely rename field names to the GeoArrow prescribed ones.

> To get back to that initial question: it will be something like `geometry.list.element.list.element.x` (with the number of `list.element` repetitions depending on the geometry type)

👍